Branch for tester

Skoteini-42 opened the pull request:
This is just a draft for the tests. Fixes are already made here for the file listing/HEAD implemented already and PUT will be done shortly in my official branch. You can do anything you want in this branch as it's just a copy from mine.

--------------------------------------------------------------------------------------------------------------------------------------------

Test the `ubuntu_tester` using the test files inside folder `tests/website/directory` (yes the tester expects us to have everything in a folder called directory, it's just shit) and use 42tester.config, make any changes you like here during your tests as well and ofc change your root.

Concerning tests 6 and 14 I had to redirect both requests to Static to see what's going on. So when you test #14, make sure to comment out `cgi .bla /home/fpapadak/webserv/tests/ubuntu_cgi_tester;` inside the location block if you want to check what's happening with my dummy Static POST. (more explanations below).

--------------------------------------------------------------------------------------------------------------------------------------------

- In order to pass the first 2 tests, we need fixes in Network - how it reads the end of raw requests. I have made a small tmp fix so the tester can move on until we have Daniel's PR.

- For the 6th test that tests a file with a CGI extension, `Test GET http://localhost:8080/directory/youpi.bla` I encounted issues. We return `500 `code + `PATH_INFO not found'`. I temporarily added this inside` _makeDecision()` in Routing, in order to redirect the GET requests to static so i can move with tests. It then returns a 200 and the tests move on, but we want the CGI to handle his in reality, not Static. So check first without `&&  _request.getMethod() == "POST"` that I added in that condition so it can pass through CGI and see the error. Then add it back to see that it succeeds from Static (but that's not correct ofc).

```
	else if (_location->isCgi(utils::getFileExtension(_request.getScriptName())) && 
             _request.getMethod() == "POST") //tmp fix!!!
```
-  For the 12th test, we return 403 with autoindex off which is correct. The tester doesn't like it, and prefers not_found, I saw that some servers actually return 404 to avoid revealing folder structures, rather than just saying "forbidden", so I changed that.

- For the 14th test, it's a real shit situation (POST test). It tries to pass a content of 100.000.000 bytes with chunked and yes, it expects a 200 OK for that! (based on numerous slack converstations I found).  I get : `make: *** [Makefile:95: test] Broken pipe` and ofc the tester returned EOF as you see below. I temp added for myself to see a dummy POST inside Static and redirected POST requests to Static just to avoid the broken pipe issue from CGI, and then the tester says `bad cgi returned size body` but the broken pipe error is solved. So, it actually wants the content-length to be 100 MB and CGI to return all this body back! It seems that some students use tmp files for large payloads, rather than piping them. Could you look into it maybe because this is again for CGI to handle - not Static (I just temp redirected it to Static to understand what it expects since there is an issue with CGI + piping). I can show you some very helpful slack converstations about that, this is the test with the biggest issue people seemed to have so far. 
--------------------------------------------------------------------------------------------------------------------------------------------

 **LIST OF ALL TESTS SO FAR :** (all pass until the last one)
 
 
1)  Test GET http://localhost:8080/
content returned: (here it returns the file listing) 

2) Test POST http://localhost:8080/ with a size of 0

3) Test HEAD http://localhost:8080/

4) Test GET http://localhost:8080/directory
(here it shows 405 error page)
content returned: 

5) Test GET http://localhost:8080/directory/youpi.bad_extension
content returned: 

6) Test GET http://localhost:8080/directory/youpi.bla ----> now it passes properly throgh CGI with new env variable added
content returned: 

7) Test GET Expected 404 on http://localhost:8080/directory/oulalala
content returned: 
(here shows 404 error page)


8) Test GET http://localhost:8080/directory/nop
content returned: 

9) Test GET http://localhost:8080/directory/nop/
content returned: 

10) Test GET http://localhost:8080/directory/nop/other.pouic
content returned: 

11) Test GET Expected 404 on http://localhost:8080/directory/nop/other.pouac
content returned: 
(here shows 404 error page)


12) Test GET Expected 404 on http://localhost:8080/directory/Yeah
content returned:
(here shows 404 error page)


13) Test GET http://localhost:8080/directory/Yeah/not_happy.bad_extension
content returned: 

14) Test POST http://localhost:8080/directory/youpi.bla with a size of 100000000
FATAL ERROR ON LAST TEST: Post "http://localhost:8080/directory/youpi.bla": EOF -> with using CGI POST
                                                                           OR
   Test POST http://localhost:8080/directory/youpi.bla with a size of 100000000
   FATAL ERROR ON LAST TEST: Post "http://localhost:8080/directory/youpi.bla": bad cgi returned size body -> with using dummy static POST, but we need to return this 100 MB content back

Skoteini-42 committed:
added dummy static back commended

edouardproust commented:
`the tester expects us to have everything in a folder called directory` -> wtf ? And how did you discovered that ??

`For the 6th test that tests a file with a CGI extension, Test GET http://localhost:8080/directory/youpi.bla I encounted issues. We return 500 code + PATH_INFO not found'` -> Yes exactly, and I have absolutly NO IDEA how to pass this...! Whatever the value we put as a PATH_INFO it is rejected by the tester

`For the 12th test, we return 403 with autoindex off which is correct. The tester doesn't like it, and prefers not_found, I saw that some servers actually return 404 to avoid revealing folder structures, rather than just saying "forbidden", so I changed that.` -> Ok fine for me!

`For the 14th test, it's a real shit situation (POST test). It tries to pass a content of 100.000.000 bytes with chunked and yes, it expects a 200 OK for that! (based on numerous slack converstations I found). I get : make: *** [Makefile:95: test] Broken pipe` -> I hope my update will fix this... Let's see

edouardproust commented:
Weird, i get this error and test#2 sometimes:

```
Test GET http://localhost:8080/
content returned:
//...

Test POST http://localhost:8080/ with a size of 0
FATAL ERROR ON LAST TEST: Post "http://localhost:8080/": read tcp 127.0.0.1:35356->127.0.0.1:8080: read: connection reset by peer
```

Skoteini-42 committed:
merged with new main

Skoteini-42 commented:
> `the tester expects us to have everything in a folder called directory` -> wtf ? And how did you discovered that ?? -> Because after the 5th test I understood it tests as if the files where inside the folder direcotry, instead of YoupiBanane as it asks. I started getting 404 errors, and once I changed the name from YoupiBanane  -> directory all seemed well and we return the status codes it expects....I don't know, that's the solution I found, maybe you have a better idea!
> 
> `For the 6th test that tests a file with a CGI extension, Test GET http://localhost:8080/directory/youpi.bla I encounted issues. We return 500 code + PATH_INFO not found'` -> Yes exactly, and I have absolutly NO IDEA how to pass this...! Whatever the value we put as a PATH_INFO it is rejected by the tester
> 
> `For the 12th test, we return 403 with autoindex off which is correct. The tester doesn't like it, and prefers not_found, I saw that some servers actually return 404 to avoid revealing folder structures, rather than just saying "forbidden", so I changed that.` -> Ok fine for me!
> 
> `For the 14th test, it's a real shit situation (POST test). It tries to pass a content of 100.000.000 bytes with chunked and yes, it expects a 200 OK for that! (based on numerous slack converstations I found). I get : make: *** [Makefile:95: test] Broken pipe` -> I hope my update will fix this... Let's see

Skoteini-42 commented:
UPDATE : Daniel's fixes are now merged here as well

Skoteini-42 commented:
> Weird, i get this error and test#2 sometimes:
> 
> ```
> Test GET http://localhost:8080/
> content returned:
> //...
> 
> Test POST http://localhost:8080/ with a size of 0
> FATAL ERROR ON LAST TEST: Post "http://localhost:8080/": read tcp 127.0.0.1:35356->127.0.0.1:8080: read: connection reset by peer
> ```

Maybe now with Daniel's updated code you won't see this anymore (merged here now), because now raw requests from both tests should be read entirely properly. I didn't see that happening at all, but if you see it let's check what's happening

Skoteini-42 committed:
merged with new main

Skoteini-42 committed:
merged with new main

Skoteini-42 committed:
cgi env fix

Skoteini-42 committed:
added debugs inside cgi io

Skoteini-42 commented:
This branch now has the new main merged + the fix for CGI env variables, still haging issues with test no14, added a few debugging lines to see what's happening
